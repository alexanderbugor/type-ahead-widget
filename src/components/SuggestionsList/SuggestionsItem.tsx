import React from 'react';
import { WithStyles, withStyles, createStyles } from '@material-ui/core/styles';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import { Suggestion } from './SuggestionsList';
import HighlightMatch from '../HighlightMatch/HighlightMatch';

const styles = createStyles({
  active: {
    backgroundColor: '#00000008',
  },
});

interface Props extends WithStyles<typeof styles> {
  item: Suggestion;
  index: number;
  activeItemIndex: number;
  onSuggestionSelect: (value: string) => void;
}

class SuggestionsItem extends React.PureComponent<Props> {
  ref = React.createRef<HTMLDivElement>();

  componentDidMount() {
    if (this.isItemActive()) {
      this.scrollItemIntoView();
    }
  }

  componentDidUpdate(prevProps: Props) {
    if (this.props.activeItemIndex !== prevProps.activeItemIndex && this.isItemActive()){
      this.scrollItemIntoView();
    }
  }

  scrollItemIntoView = () => {
    this.ref.current.scrollIntoView({
      behavior: 'smooth',
      block: 'start',
    });
  }

  isItemActive = () => {
    const { index, activeItemIndex } = this.props;

    return activeItemIndex === index;
  }
  
  render() {
    const {
      index,
      item,
      classes,
      onSuggestionSelect,
    } = this.props;

    return (
      <div ref={this.ref}>
        <ListItem
          className={this.isItemActive() ? classes.active : null}
          button
          onClick={() => onSuggestionSelect(item.item)}
        >
          <ListItemText
            primary={
              <HighlightMatch matches={item.matches}>
                {item.item}
              </HighlightMatch>
            }
          />
        </ListItem>
      </div>
    );
  }
}

export default withStyles(styles)(SuggestionsItem);
