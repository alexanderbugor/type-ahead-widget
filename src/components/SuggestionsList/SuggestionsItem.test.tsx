import React from 'react';
import { shallow } from 'enzyme';
import Fuse from 'fuse.js';
import ListItem from '@material-ui/core/ListItem';
import { SuggestionsItem } from './SuggestionsItem';

describe('SuggestionsItem', () => {
  const indices = ([1, 4] as unknown) as Fuse.RangeTuple[];
  const props = {
    item: { item: 'Banana', matches: [{ indices, value: 'Banana' }] },
    index: 2,
    activeItemIndex: 0,
    onSuggestionSelect: jest.fn(),
    classes: { active: 'MOCK_ACTIVE' },
  };

  it('should render and match snapshot', () => {
    const root = shallow<SuggestionsItem>(<SuggestionsItem {...props} />);

    expect(root).toMatchSnapshot();
  });

  it('should scroll the active item into view', () => {
    const scrollIntoView = jest.fn();
    const root = shallow<SuggestionsItem>(<SuggestionsItem {...props} />);
    root.instance().ref = {
      current: {
        scrollIntoView,
      } as unknown as HTMLDivElement,
    };
    root.setProps({
      activeItemIndex: 2,
    });

    expect(scrollIntoView).toHaveBeenCalledWith({
      behavior: 'smooth',
      block: 'start',
    });
  });

  describe('ListItem', () => {
    it('should have className when active', () => {
      const scrollIntoView = jest.fn();
      const root = shallow<SuggestionsItem>(<SuggestionsItem {...props} />);
      root.instance().ref = {
        current: {
          scrollIntoView,
        } as unknown as HTMLDivElement,
      };
      root.setProps({
        activeItemIndex: 2,
      });

      const element = root.find(ListItem);

      expect(element).toExist();
      expect(element).toHaveProp('className', props.classes.active);
    });

    it('should handleClick', () => {
      const scrollIntoView = jest.fn();
      const root = shallow<SuggestionsItem>(<SuggestionsItem {...props} />);
      root.instance().ref = {
        current: {
          scrollIntoView,
        } as unknown as HTMLDivElement,
      };
      root.setProps({
        activeItemIndex: 2,
      });

      const element = root.find(ListItem);
      element.props().onClick('MOCK_EVENT' as unknown as React.MouseEvent<HTMLDivElement>);

      expect(props.onSuggestionSelect).toHaveBeenCalledWith(props.item.item);
    });
  });
});
