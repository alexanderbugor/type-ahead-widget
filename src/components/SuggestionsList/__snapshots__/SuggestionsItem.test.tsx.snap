// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SuggestionsItem should render and match snapshot 1`] = `
<div>
  <WithStyles(ForwardRef(ListItem))
    button={true}
    className={null}
    onClick={[Function]}
  >
    <WithStyles(ForwardRef(ListItemText))
      primary={
        <HighlightMatch
          matches={
            Array [
              Object {
                "indices": Array [
                  1,
                  4,
                ],
                "value": "Banana",
              },
            ]
          }
        >
          Banana
        </HighlightMatch>
      }
    />
  </WithStyles(ForwardRef(ListItem))>
</div>
`;
